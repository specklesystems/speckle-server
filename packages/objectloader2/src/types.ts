export type CustomLogger = (message?: string, ...optionalParams: unknown[]) => void

export interface Item {
  id: string
  obj: Base
}

export interface Base {
  id: string
  __closure?: string[]
}

export function isString(value: unknown): value is string {
  return typeof value === 'string'
}
