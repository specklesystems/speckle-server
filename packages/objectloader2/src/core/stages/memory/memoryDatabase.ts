import { Base, Item } from '../../../types/types.js'
import { Database } from '../../interfaces.js'
import { MemoryDatabaseOptions } from '../../options.js'

export class MemoryDatabase implements Database {
  private items: Map<string, Base>

  constructor(options?: MemoryDatabaseOptions) {
    this.items = options?.items || new Map<string, Base>()
  }

  getAll(keys: string[]): Promise<(Item | undefined)[]> {
    const found: (Item | undefined)[] = []
    for (const key of keys) {
      const item = this.items.get(key)
      if (item) {
        found.push({ baseId: key, base: item })
      } else {
        found.push(undefined)
      }
    }
    return Promise.resolve(found)
  }

  saveBatch({ batch }: { batch: Item[] }): Promise<void> {
    for (const item of batch) {
      if (!item.baseId || !item.base) {
        throw new Error('Item must have a baseId and base')
      }
      this.items.set(item.baseId, item.base)
    }
    return Promise.resolve()
  }

  disposeAsync(): Promise<void> {
    return Promise.resolve()
  }
}
