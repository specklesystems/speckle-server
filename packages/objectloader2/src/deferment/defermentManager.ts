import { DeferredBase } from './deferredBase.js'
import { CustomLogger } from '../types/functions.js'
import { Item, Base } from '../types/types.js'
import { BaseCache } from './BaseCache.js'

export class DefermentManager {
  private deferments: Map<string, DeferredBase> = new Map()
  private logger: CustomLogger
  private disposed = false
  private cache: BaseCache

  constructor(cache: BaseCache, logger: CustomLogger) {
    this.cache = cache
    this.logger = logger
    }

    private now(): number {
      return Date.now()
    }

  get(id: string): DeferredBase | undefined {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    return this.deferments.get(id)
  }

  defer(params: { id: string }): [Promise<Base>, boolean] {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const item = this.cache.get(params.id)
    if (item) {
      return [Promise.resolve(item.base!), true]
    }
    const deferredBase = this.deferments.get(params.id)
    if (deferredBase) {
      return [deferredBase.getPromise(), true]
    }
    const notYetFound = new DeferredBase(
      params.id
    )
    this.deferments.set(params.id, notYetFound)
    return [notYetFound.getPromise(), false]
  }

  undefer(item: Item, requestItem: (id: string) => void): void {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const base = item.base
    if (!base) {
      this.logger('undefer called with no base', item)
      return
    }
    this.cache.add(item, (id) => {
      if (!this.deferments.has(id)) {
        requestItem(id)
      }
    })

    //order matters here with found before undefer
    const deferredBase = this.deferments.get(item.baseId)
    if (deferredBase) {
      deferredBase.found(base, item.size || 0)
      this.deferments.delete(item.baseId)
    }
  }

  dispose(): void {
    if (this.disposed) return
    this.disposed = true
    this.clearDeferments()
  }

  private clearDeferments(): void {
    let waiting = 0
    for (const deferredBase of this.deferments.values()) {

      if (deferredBase.getBase() === undefined) {
        waiting++
      }
    }
    this.deferments.clear()
    this.logger('cleared deferments, left', waiting)
  }
}
