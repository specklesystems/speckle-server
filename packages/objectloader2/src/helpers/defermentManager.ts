import { DeferredBase } from './deferredBase.js'
import { Base, CustomLogger, Item } from '../types/types.js'
import { DefermentManagerOptions } from '../operations/options.js'

export class DefermentManager {
  private deferments: Map<string, DeferredBase> = new Map()
  private timer?: ReturnType<typeof setTimeout>
  private logger: CustomLogger
  private currentSize = 0
  private disposed = false

  constructor(private options: DefermentManagerOptions) {
    this.resetGlobalTimer()
    this.logger = options.logger || ((): void => {})
  }

  private now(): number {
    return Date.now()
  }

  isDeferred(id: string): boolean {
    return this.deferments.has(id)
  }

  get(id: string): DeferredBase | undefined {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    return this.deferments.get(id)
  }

  async defer(params: { id: string }): Promise<Base> {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const now = this.now()
    const deferredBase = this.deferments.get(params.id)
    if (deferredBase) {
      deferredBase.setAccess(now)
      return deferredBase.getPromise()
    }
    const notYetFound = new DeferredBase(
      this.options.ttlms,
      params.id,
      now + this.options.ttlms
    )
    this.deferments.set(params.id, notYetFound)
    return notYetFound.getPromise()
  }

  undefer(item: Item): void {
    if (this.disposed) throw new Error('DefermentManager is disposed')
    const now = this.now()
    this.currentSize += item.size || 0
    //order matters here with found before undefer
    const deferredBase = this.deferments.get(item.baseId)
    if (deferredBase) {
      deferredBase.found(item)
      deferredBase.setAccess(now)
    } else {
      const existing = new DeferredBase(this.options.ttlms, item.baseId, now)
      existing.found(item)
      this.deferments.set(item.baseId, existing)
    }
  }

  private resetGlobalTimer(): void {
    const run = (): void => {
      this.cleanDeferments()
      this.timer = setTimeout(run, this.options.ttlms)
    }
    this.timer = setTimeout(run, this.options.ttlms)
  }

  dispose(): void {
    if (this.disposed) return
    this.disposed = true
    if (this.timer) {
      clearTimeout(this.timer)
      this.timer = undefined
    }
    this.clearDeferments()
  }

  private clearDeferments(): void {
    let waiting = 0
    for (const deferredBase of this.deferments.values()) {
      deferredBase.done(0)
      if (deferredBase.getItem() === undefined) {
        waiting++
      }
    }
    this.currentSize = 0
    this.deferments.clear()
    this.logger('cleared deferments, left', waiting)
  }

  private cleanDeferments(): void {
    const maxSizeBytes = this.options.maxSizeInMb * 1024 * 1024
    if (this.currentSize < maxSizeBytes) {
      this.logger(
        'deferments size is ok, no need to clean',
        this.currentSize,
        maxSizeBytes
      )
      return
    }
    const now = this.now()
    let cleaned = 0
    const start = performance.now()
    for (const deferredBase of Array.from(this.deferments.values())
      .filter((x) => x.isExpired(now))
      .sort((a, b) => this.compareMaybeBasesBySize(a.getItem(), b.getItem()))) {
      if (deferredBase.done(now)) {
        this.currentSize -= deferredBase.getItem()?.size || 0
        this.deferments.delete(deferredBase.getId())
        cleaned++
        if (this.currentSize < maxSizeBytes) {
          break
        }
      }
    }
    this.logger(
      'cleaned deferments, cleaned, left',
      cleaned,
      this.deferments.size,
      performance.now() - start
    )
    return
  }

  compareMaybeBasesBySize(a: Item | undefined, b: Item | undefined): number {
    if (a === undefined && b === undefined) return 0
    if (a === undefined) return -1
    if (b === undefined) return 1
    return this.compareMaybe(a.size, b.size)
  }

  compareMaybe(a: number | undefined, b: number | undefined): number {
    if (a === undefined && b === undefined) return 0
    if (a === undefined) return -1
    if (b === undefined) return 1
    return a - b
  }
}
