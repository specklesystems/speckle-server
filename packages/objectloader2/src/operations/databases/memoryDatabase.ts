import { Base, Item } from '../../types/types.js'
import { Database } from '../interfaces.js'
import { MemoryDatabaseOptions } from '../options.js'

export class MemoryDatabase implements Database {
  private items: Map<string, Base>

  constructor(options?: MemoryDatabaseOptions) {
    this.items = options?.items || new Map<string, Base>()
  }

  getAll(keys: string[]): Promise<(Item | undefined)[]> {
    const found: (Item | undefined)[] = []
    for (const key of keys) {
      const item = this.items.get(key)
      if (item) {
        found.push({ baseId: key, base: item })
      } else {
        found.push(undefined)
      }
    }
    return Promise.resolve(found)
  }

  cacheSaveBatch({ batch }: { batch: Item[] }): Promise<void> {
    for (const item of batch) {
      this.items.set(item.baseId, item.base)
    }
    return Promise.resolve()
  }

  getItem(params: { id: string }): Promise<Item | undefined> {
    const item = this.items.get(params.id)
    if (item) {
      return Promise.resolve({ baseId: params.id, base: item })
    }
    return Promise.resolve(undefined)
  }

  disposeAsync(): Promise<void> {
    return Promise.resolve()
  }
}
