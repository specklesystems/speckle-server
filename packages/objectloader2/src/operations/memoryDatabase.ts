import Queue from '../helpers/queue.js'
import { Base, Item } from '../types/types.js'
import { Cache } from './interfaces.js'
import { MemoryDatabaseOptions } from './options.js'

export class MemoryDatabase implements Cache {
  #items: Record<string, Base>
  constructor(options?: MemoryDatabaseOptions) {
    this.#items = options?.items || {}
  }

  getItem(params: { id: string }): Promise<Item | undefined> {
    const item = this.#items[params.id]
    if (item) {
      return Promise.resolve({ baseId: params.id, base: item })
    }
    return Promise.resolve(undefined)
  }
  processItems(params: {
    ids: string[]
    foundItems: Queue<Item>
    notFoundItems: Queue<string>
  }): Promise<void> {
    const { ids, foundItems, notFoundItems } = params
    for (const id of ids) {
      const item = this.#items[id]
      if (item) {
        foundItems.add({ baseId: id, base: item })
      } else {
        notFoundItems.add(id)
      }
    }
    return Promise.resolve()
  }
  add(item: Item): Promise<void> {
    this.#items[item.baseId] = item.base
    return Promise.resolve()
  }
  disposeAsync(): Promise<void> {
    return Promise.resolve()
  }
}
