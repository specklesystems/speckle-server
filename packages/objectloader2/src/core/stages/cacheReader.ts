import { Deferment } from '../../deferment/defermentManager.js'
import BatchingQueue from '../../queues/batchingQueue.js'
import Queue from '../../queues/queue.js'
import { CustomLogger } from '../../types/functions.js'
import { Item, Base } from '../../types/types.js'
import { Database } from '../interfaces.js'
import { CacheOptions } from '../options.js'

export class CacheReader {
  #database: Database
  #defermentManager: Deferment
  #logger: CustomLogger
  #options: CacheOptions
  #readQueue: BatchingQueue<string> | undefined
  #foundQueue: Queue<Item> | undefined
  #notFoundQueue: Queue<string> | undefined

  constructor(
    database: Database,
    defermentManager: Deferment,
    logger: CustomLogger,
    options: CacheOptions
  ) {
    this.#database = database
    this.#defermentManager = defermentManager
    this.#logger = logger
    this.#options = options
  }

  initializeQueue(foundQueue: Queue<Item>, notFoundQueue: Queue<string>): void {
    this.#foundQueue = foundQueue
    this.#notFoundQueue = notFoundQueue
  }

  getObject(params: { id: string }): Promise<Base> {
    const [p, b] = this.#defermentManager.defer({ id: params.id })
    if (!b) {
      this.requestItem(params.id)
    }
    return p
  }

  #createReadQueue(): void {
    if (!this.#readQueue) {
      this.#readQueue = new BatchingQueue({
        batchSize: this.#options.maxCacheReadSize,
        maxWaitTime: this.#options.maxCacheBatchReadWait,
        processFunction: this.#processBatch
      })
    }
  }

  requestItem(id: string): void {
    this.#createReadQueue()
    if (!this.#readQueue?.get(id)) {
      this.#readQueue?.add(id, id)
    }
  }

  requestAll(keys: string[]): void {
    this.#createReadQueue()

    this.#readQueue?.addAll(keys, keys)
  }

  #processBatch = async (batch: string[]): Promise<void> => {
    const start = performance.now()
    const items = await this.#database.getAll(batch)
    for (let i = 0; i < items.length; i++) {
      const item = items[i]
      if (item) {
        this.#foundQueue?.add(item)
        this.#defermentManager.undefer(item, (id) => this.requestItem(id))
      } else {
        this.#notFoundQueue?.add(batch[i])
      }
    }
    this.#logger(
      `readBatch: batch ${batch.length}, time ${
        performance.now() - start
      } ms left ${this.#readQueue?.count()}`
    )
  }

  disposeAsync(): Promise<void> {
    return this.#readQueue?.disposeAsync() || Promise.resolve()
  }
}
