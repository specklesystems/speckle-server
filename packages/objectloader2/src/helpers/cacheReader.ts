import { Database } from '../operations/interfaces.js'
import { CacheOptions } from '../operations/options.js'
import { Base, CustomLogger, Item } from '../types/types.js'
import BatchingQueue from './batchingQueue.js'
import { DefermentManager } from './defermentManager.js'

export class CacheReader {
  #database: Database
  #defermentManager: DefermentManager
  #logger: CustomLogger
  #options: CacheOptions
  #readQueue: BatchingQueue<string> | undefined

  constructor(
    database: Database,
    defermentManager: DefermentManager,
    options: CacheOptions
  ) {
    this.#database = database
    this.#defermentManager = defermentManager
    this.#options = options
    this.#logger = options.logger || ((): void => {})
  }

  async getObject(params: { id: string }): Promise<Base> {
    if (!this.#defermentManager.isDeferred(params.id)) {
      this.#getItem(params.id)
    }
    return await this.#defermentManager.defer({ id: params.id })
  }

  #getItem(id: string): void {
    if (!this.#readQueue) {
      this.#readQueue = new BatchingQueue({
        batchSize: this.#options.maxCacheReadSize,
        maxWaitTime: this.#options.maxCacheBatchReadWait,
        processFunction: this.#processBatch
      })
    }
    if (!this.#readQueue.get(id)) {
      this.#readQueue.add(id, id)
    }
  }

  async getAll(keys: string[]): Promise<(Item | undefined)[]> {
    return this.#database.getAll(keys)
  }

  #processBatch = async (batch: string[]): Promise<void> => {
    const items = await this.#database.getAll(batch)
    for (let i = 0; i < items.length; i++) {
      if (items[i]) {
        this.#defermentManager.undefer(items[i]!)
      } else {
        //this is okay!
        //this.#logger(`Item ${batch[i]} not found in cache`)
      }
    }
  }

  async disposeAsync(): Promise<void> {
    await this.#readQueue?.disposeAsync()
  }
}
