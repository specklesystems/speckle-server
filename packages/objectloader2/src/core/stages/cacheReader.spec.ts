import { describe, test, expect } from 'vitest'
import { DefermentManager } from '../../deferment/defermentManager.js'
import { Item, Base } from '../../types/types.js'
import { CacheReader } from './cacheReader.js'
import { MemoryDatabase } from './memory/memoryDatabase.js'
import { MemoryCache } from '../../deferment/MemoryCache.js'

describe('CacheReader testing', () => {
  test('deferred getObject', async () => {
    const i1: Item = { baseId: 'id1', base: { id: 'id', speckle_type: 'type' } }

    const deferments = new DefermentManager(
      new MemoryCache({ maxSizeInMb: 1, ttlms: 1 }, () => {}),
      () => {}
    )
    const cacheReader = new CacheReader(
      new MemoryDatabase({
        items: new Map<string, Base>([[i1.baseId, i1.base!]])
      }),
      deferments,
      {
        maxCacheReadSize: 1,
        maxCacheWriteSize: 1,
        maxCacheBatchWriteWait: 1,
        maxCacheBatchReadWait: 1,
        maxWriteQueueSize: 1
      }
    )

    const objPromise = cacheReader.getObject({
      id: i1.baseId
    })
    deferments.undefer(i1, (id: string) => cacheReader.requestItem(id))
    const base = await objPromise

    expect(base).toMatchSnapshot()
    await cacheReader.disposeAsync()
  })
})
