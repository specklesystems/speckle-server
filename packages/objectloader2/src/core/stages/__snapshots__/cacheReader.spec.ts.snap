// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CacheReader testing > deferred getObject 1`] = `
{
  "id": "id",
  "speckle_type": "type",
}
`;
